### PR TITLE
feat: add telemetry for task lifecycle events

### PR DESCRIPTION
**Pull request by johnstcn**
Relates to https://github.com/coder/internal/issues/1259

Adds new database queries and telemetry collection functions to gather task lifecycle events (pause/resume cycles, idle time) for analytics.
    
Task events track pause/resume activity, idle duration before pausing, paused duration, and time from resume to first app status, filtered to recent activity based on the telemetry snapshot interval.

🤖 Created with Mux (Opus 4.6).

### CODE REVIEW

File: coderd/database/queries/workspacebuilds.sql [hunk 1]
```diff
@@ -21,6 +21,39 @@ LIMIT
...
+    (
+        SELECT id, workspace_id, created_at, transition, reason
+        FROM workspace_builds
+        WHERE workspace_id = ws.id
+          AND transition = 'start'
+        ORDER BY created_at DESC
+        LIMIT 1
+    )
+) b
+ORDER BY b.workspace_id, b.created_at DESC;
```
Comment (johnstcn) [inline coderd/database/queries/workspacebuilds.sql]: Sample query on dogfood db (no index): https://explain.dalibo.com/plan/ad36577g512863cb

File: coderd/telemetry/telemetry.go [hunk 2]
```diff
@@ -902,6 +904,105 @@ func (r *remoteReporter) collectBoundaryUsageSummary(ctx context.Context) (*Boun
...
+	}
+
+	// Resume-to-status: time from the latest start build to the
+	// first app status reported after that start.
+	if running && row.FirstStatusAfterResumeAt.Valid {
+		delta := row.FirstStatusAfterResumeAt.Time.Sub(row.StartBuildCreatedAt.Time)
+		event.ResumeToStatusMS = ptr.Ref(delta.Milliseconds())
+	}
+
+	return event
```
Comment (johnstcn) [inline coderd/telemetry/telemetry.go]: self-review: might make sense to re-order and combine some of these (e.g. L958, L971).

File: coderd/telemetry/telemetry_test.go [hunk 3]
```diff
@@ -313,6 +317,15 @@ func TestTelemetry(t *testing.T) {
...
 		require.Len(t, snapshot.Tasks, 1)
+		require.Len(t, snapshot.TaskEvents, 1)
+		taskEvent := snapshot.TaskEvents[0]
+		assert.Equal(t, task.ID.String(), taskEvent.TaskID)
+		assert.Nil(t, taskEvent.LastResumedAt)
+		assert.Nil(t, taskEvent.LastPausedAt)
+		assert.Nil(t, taskEvent.PauseReason)
+		assert.Nil(t, taskEvent.IdleDurationMS)
+		assert.Nil(t, taskEvent.PausedDurationMS)
+		assert.Nil(t, taskEvent.ResumeToStatusMS)
```
Comment (johnstcn) [inline coderd/telemetry/telemetry_test.go:329]: self-review: adding some task app statuses here may not be a bad idea.

File: coderd/database/queries/tasks.sql [hunk 4]
```diff
@@ -100,3 +100,76 @@ FROM
...
+        WHERE was.app_id = twa.workspace_app_id
+          AND was.state = 'working'
+        ORDER BY was.created_at DESC
+        LIMIT 1
+    ) lws ON twa.workspace_app_id IS NOT NULL
+    LEFT JOIN LATERAL (
+        SELECT was.created_at
+        FROM workspace_app_statuses was
+        WHERE was.app_id = twa.workspace_app_id
+          AND was.created_at > start_build.created_at
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: Suggestion: Prefer standard indentation against alignment. (Same elsewhere in the query, including select). Doing so makes editing easier.

File: coderd/database/queries/tasks.sql [hunk 5]
```diff
@@ -100,3 +100,76 @@ FROM
...
+WITH task_event_data AS (
+    SELECT
+        t.id AS task_id,
+        t.workspace_id,
+        twa.workspace_app_id,
+        -- Latest stop build.
+        stop_build.created_at   AS stop_build_created_at,
+        stop_build.reason       AS stop_build_reason,
+        -- Latest start build.
+        start_build.created_at  AS start_build_created_at,
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: Suggestion: Missing from ticket, but consider using reason here too (e.g. task resume). Granted, the only one we have currently is manual resume, if it's missing it was likely a manual workspace start or workspace auto start schedule.

File: coderd/database/queries/tasks.sql [hunk 6]
```diff
@@ -100,3 +100,76 @@ FROM
...
+        ORDER BY wb.build_number DESC
+        LIMIT 1
+    ) start_build ON TRUE
+    LEFT JOIN LATERAL (
+        SELECT was.created_at
+        FROM workspace_app_statuses was
+        WHERE was.app_id = twa.workspace_app_id
+          AND was.state = 'working'
+        ORDER BY was.created_at DESC
+        LIMIT 1
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: Nit: This is something of an approximation. Consider a situation where the agent reports working, goes on to do 1h of work, and then reports done. We'd be missing 1h of work from this timeframe.

We could improve this, but I'm not sure it's worth the lift. But a comment would be good IMO.

File: coderd/database/queries/tasks.sql [hunk 7]
```diff
@@ -100,3 +100,76 @@ FROM
...
+        -- Latest stop build.
+        stop_build.created_at   AS stop_build_created_at,
+        stop_build.reason       AS stop_build_reason,
+        -- Latest start build.
+        start_build.created_at  AS start_build_created_at,
+        -- Last "working" app status (for idle duration).
+        lws.created_at          AS last_working_status_at,
+        -- First app status after resume (for resume-to-status duration).
+        -- Only populated for workspaces in an active phase (started more
+        -- recently than stopped).
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: Was this in the ticket/RFC? Either way, this seems like a useful metric to add. 👍🏻

File: coderd/telemetry/telemetry.go [hunk 8]
```diff
@@ -902,6 +904,105 @@ func (r *remoteReporter) collectBoundaryUsageSummary(ctx context.Context) (*Boun
...
+			event.PauseReason = ptr.Ref("other")
+		}
+	}
+
+	// Last resumed (latest start build that follows a stop build).
+	if startedAfterStop {
+		event.LastResumedAt = &row.StartBuildCreatedAt.Time
+	}
+
+	// Idle duration: time between last working status and latest stop.
```
Comment (mafredri) [inline coderd/telemetry/telemetry.go]: Question: Should we include running tasks as well? They will have a constantly increasing  idle duration and may be hidden from the telemetry otherwise.

File: coderd/telemetry/telemetry.go [hunk 9]
```diff
@@ -902,6 +904,105 @@ func (r *remoteReporter) collectBoundaryUsageSummary(ctx context.Context) (*Boun
...
+		event.IdleDurationMS = ptr.Ref(idle.Milliseconds())
+	}
+
+	// Paused duration: time between latest stop and latest start,
+	// gated by createdAfter so only recent pairs are reported.
+	if startedAfterStop &&
+		row.StartBuildCreatedAt.Time.After(createdAfter) {
+		paused := row.StartBuildCreatedAt.Time.Sub(row.StopBuildCreatedAt.Time)
+		event.PausedDurationMS = ptr.Ref(paused.Milliseconds())
+	}
```
Comment (mafredri) [inline coderd/telemetry/telemetry.go:993]: Question: Similar to the comment above, are we interested in an ever-increasing paused duration being shown for tasks that don't get resumed?

File: coderd/database/queries/tasks.sql [hunk 10]
```diff
@@ -100,3 +100,76 @@ FROM
...
+        t.id AS task_id,
+        t.workspace_id,
+        twa.workspace_app_id,
+        -- Latest stop build.
+        stop_build.created_at   AS stop_build_created_at,
+        stop_build.reason       AS stop_build_reason,
+        -- Latest start build.
+        start_build.created_at  AS start_build_created_at,
+        -- Last "working" app status (for idle duration).
+        lws.created_at          AS last_working_status_at,
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: This was lost in the RFC -> ticket conversion, but we also want a "active duration", i.e. how much time the task spent working.

File: coderd/telemetry/telemetry.go [hunk 11]
```diff
@@ -902,6 +904,105 @@ func (r *remoteReporter) collectBoundaryUsageSummary(ctx context.Context) (*Boun
...
+		}
+	}
+
+	// Last resumed (latest start build that follows a stop build).
+	if startedAfterStop {
+		event.LastResumedAt = &row.StartBuildCreatedAt.Time
+	}
+
+	// Idle duration: time between last working status and latest stop.
+	if hasStopBuild && row.LastWorkingStatusAt.Valid &&
```
Comment (mafredri) [inline coderd/telemetry/telemetry.go]: Suggestion: Consider grouping the hasStopBuild and startedAfterStop code branches.

File: coderd/telemetry/telemetry.go [hunk 12]
```diff
@@ -902,6 +904,105 @@ func (r *remoteReporter) collectBoundaryUsageSummary(ctx context.Context) (*Boun
...
+	// gated by createdAfter so only recent pairs are reported.
+	if startedAfterStop &&
+		row.StartBuildCreatedAt.Time.After(createdAfter) {
+		paused := row.StartBuildCreatedAt.Time.Sub(row.StopBuildCreatedAt.Time)
+		event.PausedDurationMS = ptr.Ref(paused.Milliseconds())
+	}
+
+	// Resume-to-status: time from the latest start build to the
+	// first app status reported after that start.
+	if running && row.FirstStatusAfterResumeAt.Valid {
```
Comment (mafredri) [inline coderd/telemetry/telemetry.go]: It seems like we're redoing a lot of the conditionals from the query in the code (above too). Isn't the query set up such that running is already checked and we can just rely on `.Valid` here?

**Review by mafredri — COMMENTED** [review]
Nice work!

Some small suggestions, little cleanup, and consolidating query/code conditionals, but otherwise this is looking good.

File: coderd/database/queries/tasks.sql (hunk 7 above)
Comment (johnstcn) [reply coderd/database/queries/tasks.sql]: Nope, it was just a drive-by idea I mentioned briefly in Slack 👍

File: coderd/telemetry/telemetry.go (hunk 8 above)
Comment (johnstcn) [reply coderd/telemetry/telemetry.go]: We report the task status from all tasks so I think we can leave out idle duration for running tasks. Also I think a constantly increasing idle duration for running task would be confusing.

File: coderd/telemetry/telemetry.go (hunk 9 above)
Comment (johnstcn) [reply coderd/telemetry/telemetry.go:993]: This may make sense to include.

File: coderd/database/queries/tasks.sql [hunk 13]
```diff
@@ -100,3 +100,112 @@ FROM
...
+            WHERE was.app_id = twa.workspace_app_id
+        ) intervals
+        WHERE intervals.state = 'working'
+    ) active_dur ON twa.workspace_app_id IS NOT NULL
+    WHERE t.deleted_at IS NULL
+        AND t.workspace_id IS NOT NULL
+        AND EXISTS (
+            SELECT 1 FROM workspace_builds wb
+            WHERE wb.workspace_id = t.workspace_id
+              AND wb.created_at > @created_after
```
Comment (johnstcn) [inline coderd/database/queries/tasks.sql]: review: now filtering by `created_after`

File: coderd/telemetry/telemetry.go (hunk 9 above)
Comment (johnstcn) [reply coderd/telemetry/telemetry.go:993]: b3371cb

File: coderd/database/queries/tasks.sql (hunk 6 above)
Comment (johnstcn) [reply coderd/database/queries/tasks.sql]: Comment added above:
```sql
-- - Idle duration approximation: If the agent reports "working", does
--   work, then reports "done", we miss that working time.
```

File: coderd/database/queries/tasks.sql [hunk 14]
```diff
@@ -100,3 +100,112 @@ FROM
...
+            AND wb.transition = 'stop'
+        ORDER BY wb.build_number DESC
+        LIMIT 1
+    ) stop_build ON TRUE
+    LEFT JOIN LATERAL (
+        SELECT wb.created_at
+        FROM workspace_builds wb
+        WHERE wb.workspace_id = t.workspace_id
+            AND wb.transition = 'start'
+            AND wb.reason = 'task_resume'
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: I think we should, rather than filter on reason here, select the reason in the query and add task_resume/other in the telemetry, similar to stop.

If a user manually starts the task workspace, that's still a task resumption, but not the expected way (valuable in telemetry IMO).

File: coderd/database/queries/tasks.sql [hunk 15]
```diff
@@ -100,3 +100,112 @@ FROM
...
+            OR start_build.created_at > stop_build.created_at)
+    -- Active duration: cumulative time spent in "working" state.
+    -- Uses LEAD() to convert point-in-time statuses into intervals, then sums
+    -- intervals where state='working'. For the last status, falls back to
+    -- stop_build time (if paused) or @now (if still running).
+    LEFT JOIN LATERAL (
+        SELECT COALESCE(
+            SUM(EXTRACT(EPOCH FROM (interval_end - interval_start)) * 1000)::bigint,
+            0
+        ) AS total_working_ms
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]:
```suggestion
        )::bigint AS total_working_ms
```
Go type seems to be `interface{}`, I think this should fix it (preferably here, but if that doesn't work, move it to the top `SELECT`).

File: coderd/database/queries/tasks.sql [hunk 16]
```diff
@@ -100,3 +100,112 @@ FROM
...
+        FROM (
+            SELECT
+                was.created_at AS interval_start,
+                COALESCE(
+                    LEAD(was.created_at) OVER (ORDER BY was.created_at ASC),
+                    COALESCE(stop_build.created_at, @now::timestamptz)
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: I think this might need a rethink.

We actually know which build twa.workspace_app_id belongs to, we just don't select the build number from that table.

Also, a stop build (theoretically) never has `was` content. So these can only come from the start build (i.e. task is running). For a running task, `was.created_at` will always be newer than `stop_build.created_at`. So if we ever hit this condition, we'll get interval_start > interval_end.

_By the time the task gets paused, the twa app will have changed and we will not get any data attached._

So we get a pause event, but no indication how long the task was active.

One way we could improve the data and make it more explicit is to select only the start_builds app ID from twa, and use that in our joins. That would always give us the relevant data.

File: coderd/telemetry/telemetry_test.go [hunk 17]
```diff
@@ -675,6 +689,507 @@ func TestPrebuiltWorkspacesTelemetry(t *testing.T) {
...
+		JobID:             job.ID,
+		Transition:        transition,
+		Reason:            reason,
+		BuildNumber:       buildNumber,
+		CreatedAt:         createdAt,
+		HasAITask: sql.NullBool{
+			Bool:  true,
+			Valid: true,
+		},
+	})
```
Comment (mafredri) [inline coderd/telemetry/telemetry_test.go:730]: We're missing `UpsertTaskWorkspaceApp`, which might be why the test isn't capturing what I raised on the query.

(For start, we need to sett the build/agent/app, for stop just the build).

File: coderd/database/queries/tasks.sql (hunk 15 above)
Comment (johnstcn) [reply coderd/database/queries/tasks.sql]: That did the trick 👍

File: coderd/database/queries/tasks.sql [hunk 18]
```diff
@@ -100,3 +100,147 @@ FROM
...
+	FROM workspace_app_statuses was
+	JOIN task_app_ids tai ON tai.workspace_app_id = was.app_id
+	UNION ALL
+	SELECT t.id AS task_id, wb.created_at, '_boundary' AS state
+	FROM tasks t
+	JOIN workspace_builds wb ON wb.workspace_id = t.workspace_id
+	WHERE t.deleted_at IS NULL
+		AND t.workspace_id IS NOT NULL
+		AND wb.build_number > 1
+),
```
Comment (johnstcn) [inline coderd/database/queries/tasks.sql:139]: review: extracted this into a separate CTE so we can also use this for `last_working_status`.

File: coderd/telemetry/telemetry_test.go [hunk 19]
```diff
@@ -675,6 +690,545 @@ func TestPrebuiltWorkspacesTelemetry(t *testing.T) {
...
+			resumeToStatusMS:  ptr.Ref((5 * time.Minute).Milliseconds()),
+			// Build 1 ("started work") -> Build 2 (stop) (5h10m) + Build 3 ("resumed work") -> now (25m)
+			// TODO(cian): We define IdleDurationMS as "the time from the last working status to pause".
+			//     We know that the task has reported working since T-6h and got auto-paused at T-50m.
+			//     We can reasonably assume that it has been 'idle' from when it was stopped (T-30m) to
+			//     its next report at T-25m. This is covered by ResumeToStatusMS.
+			//     But do we consider the time since its last report (T-6h) to its being auto-paused
+			//     as truly "idle"?
+			idleDurationMS:   ptr.Ref(310 * time.Minute.Milliseconds()),
+			activeDurationMS: ptr.Ref((5*time.Hour + 10*time.Minute + 25*time.Minute).Milliseconds()),
```
Comment (johnstcn) [inline coderd/telemetry/telemetry_test.go:967]: Review: AFAICT the test as written is "correct" as per the definitions laid out in the OG issue. I had some doubts about the definition but my brain may have encountered E_INSUFFICIENT_TOKENS.

File: coderd/database/sqlc.yaml [hunk 20]
```diff
@@ -124,6 +124,21 @@ sql:
...
               type: "NullWorkspaceAppHealth"
+          # Workaround for sqlc not interpreting the left join correctly
+          # in the combined telemetry query.
+          - column: "task_event_data.start_build_number"
+            go_type: "database/sql.NullInt32"
+          - column: "task_event_data.stop_build_created_at"
+            go_type: "database/sql.NullTime"
+          - column: "task_event_data.stop_build_reason"
+            go_type:
+              type: "NullBuildReason"
```
Comment (mafredri) [inline coderd/database/sqlc.yaml:135]: To cover our bases, we should include `start_build_reason` here as well (left join). Although technically there is currently always a build nr. 1 so it should never happen.

File: coderd/database/queries/tasks.sql [hunk 21]
```diff
@@ -100,3 +100,146 @@ FROM
...
+	SELECT tai.task_id, was.created_at, was.state::text AS state
+	FROM workspace_app_statuses was
+	JOIN task_app_ids tai ON tai.workspace_app_id = was.app_id
+	UNION ALL
+	SELECT t.id AS task_id, wb.created_at, '_boundary' AS state
+	FROM tasks t
+	JOIN workspace_builds wb ON wb.workspace_id = t.workspace_id
+	WHERE t.deleted_at IS NULL
+		AND t.workspace_id IS NOT NULL
+		AND wb.build_number > 1
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql:138]: Pseudo: `AND was.created_at <= MAX(wb.created_at)`?

This would be relevant for the `LEAD` / `OVER`.

File: coderd/database/queries/tasks.sql [hunk 22]
```diff
@@ -100,3 +100,146 @@ FROM
...
+    LEFT JOIN LATERAL (
+        SELECT COALESCE(
+            SUM(EXTRACT(EPOCH FROM (interval_end - interval_start)) * 1000)::bigint,
+            0
+        )::bigint AS total_working_ms
+        FROM (
+            SELECT
+                tst.created_at AS interval_start,
+                COALESCE(
+                    LEAD(tst.created_at) OVER (ORDER BY tst.created_at ASC),
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]: Edge case: Related to above comment about filtering `was.created_at`, if `was.created_at` is exactly `wb.created_at`, ordering (and thus boundary) is not guaranteed.

```sql
LEAD(tst.created_at) OVER (
	ORDER BY tst.created_at ASC,
	CASE WHEN tst.state = '_boundary' THEN 1 ELSE 0 END ASC
)
```

Also: Noticing mixed space/tab here in GH? => convert to tab?

File: coderd/database/queries/tasks.sql [hunk 23]
```diff
@@ -100,3 +100,146 @@ FROM
...
+--   work, then reports "done", we miss that working time.
+-- - lws and active_dur join across all historical app IDs for the task,
+--   because each resume cycle provisions a new app ID. This ensures
+--   pre-pause statuses contribute to idle duration and active duration.
+WITH task_app_ids AS (
+    SELECT task_id, workspace_app_id
+    FROM task_workspace_apps
+),
+task_status_timeline AS (
+	-- All app statusese across every historical app for each task,
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]:
```suggestion
	-- All app statuses across every historical app for each task,
```

File: coderd/database/queries/tasks.sql [hunk 24]
```diff
@@ -100,3 +100,146 @@ FROM
...
+	-- plus synthectic "boundary" rows at each stop/start build transition.
```
Comment (mafredri) [inline coderd/database/queries/tasks.sql]:
```suggestion
	-- plus synthetic "boundary" rows at each stop/start build transition.
```

File: coderd/telemetry/telemetry_test.go [hunk 25]
```diff
@@ -675,6 +690,545 @@ func TestPrebuiltWorkspacesTelemetry(t *testing.T) {
...
+				{2, -60 * time.Minute, database.WorkspaceTransitionStop, database.BuildReasonTaskAutoPause, nil},
+				{3, -30 * time.Minute, database.WorkspaceTransitionStart, database.BuildReasonInitiator, nil},
+			},
+			expectEvent:       true,
+			lastPausedOffset:  ptr.Ref(-60 * time.Minute),
+			lastResumedOffset: ptr.Ref(-30 * time.Minute),
+			pauseReason:       ptr.Ref("auto"),
+			resumeReason:      ptr.Ref("other"),
+			// LastResumedAt is nil because start wasn't task_resume.
+			// PausedDurationMS reports ongoing pause: now - (-60min) = 60min.
```
Comment (mafredri) [inline coderd/telemetry/telemetry_test.go]:
```suggestion
			// LastResumedAt is set because isResumed is true (build_number > 1),
			// even though the start reason isn't task_resume.
			// PausedDurationMS is the completed pause: start(-30min) - stop(-60min) = 30min.

```
Claude caught this one, seems accurate?

File: coderd/telemetry/telemetry_test.go [hunk 26]
```diff
@@ -313,6 +317,16 @@ func TestTelemetry(t *testing.T) {
...
+		require.Len(t, snapshot.TaskEvents, 1)
+		taskEvent := snapshot.TaskEvents[0]
+		assert.Equal(t, task.ID.String(), taskEvent.TaskID)
+		assert.Nil(t, taskEvent.LastResumedAt)
+		assert.Nil(t, taskEvent.LastPausedAt)
+		assert.Nil(t, taskEvent.PauseReason)
+		assert.Nil(t, taskEvent.ResumeReason)
+		assert.Nil(t, taskEvent.IdleDurationMS)
+		assert.Nil(t, taskEvent.PausedDurationMS)
+		assert.Nil(t, taskEvent.ResumeToStatusMS)
```
Comment (mafredri) [inline coderd/telemetry/telemetry_test.go:329]:
```suggestion
		assert.Nil(t, taskEvent.ResumeToStatusMS)
		assert.Nil(t, taskEvent.ActiveDurationMS)
```
Claude caught this: Everything else seems asserted?

**Review by mafredri — APPROVED** [review]
Didn't see any major issues here, nice work on the query and test coverage!

I did call out an edge case that I'd recommend addressing just to avoid any funky data (a test-case to reproduce it might be good to see what happens).

Other than that, minor nits that I don't need to re-review.

File: coderd/database/queries/tasks.sql (hunk 21 above)
Comment (johnstcn) [reply coderd/database/queries/tasks.sql:138]: It may be easier to filter down in the `lws` lateral join.

File: coderd/database/sqlc.yaml (hunk 20 above)
Comment (johnstcn) [reply coderd/database/sqlc.yaml:135]: 47dfeef

File: coderd/database/queries/tasks.sql (hunk 22 above)
Comment (johnstcn) [reply coderd/database/queries/tasks.sql]: 72adb6c

File: coderd/telemetry/telemetry_test.go (hunk 26 above)
Comment (johnstcn) [reply coderd/telemetry/telemetry_test.go:329]: e1cf392

File: coderd/telemetry/telemetry_test.go (hunk 25 above)
Comment (johnstcn) [reply coderd/telemetry/telemetry_test.go]: https://github.com/coder/coder/commit/e1cf39201c7a0fc2e65c29cbccc72a038262207d

File: coderd/database/queries/tasks.sql (hunk 24 above)
Comment (johnstcn) [reply coderd/database/queries/tasks.sql]: https://github.com/coder/coder/commit/e1cf39201c7a0fc2e65c29cbccc72a038262207d

File: coderd/database/queries/tasks.sql (hunk 23 above)
Comment (johnstcn) [reply coderd/database/queries/tasks.sql]: https://github.com/coder/coder/commit/e1cf39201c7a0fc2e65c29cbccc72a038262207d
